sdk/samples/iot/provisioning/src/paho_iot_sas_provisioning_example.c

```diff
@@ -0,0 +1,617 @@
+    char* default_value,
+    bool hide_value,
+    az_span buffer,
+    az_span* out_value);
+static void generate_sas_key();
+static uint32_t get_epoch_expiration_time_from_hours(uint32_t hours);
+static void connect_client_to_provisioning_service();
+static void subscribe_client_to_provisioning_service_topics();
+static void register_client_with_provisioning_service();
+static void receive_registration_status();
+static void parse_operation_message(
+    char* topic,
+    int topic_len,
+    const MQTTClient_message* message,
+    az_iot_provisioning_client_register_response* response,
+    az_iot_provisioning_client_operation_status* operation_status);
+static void send_operation_query_message(
+    const az_iot_provisioning_client_register_response* response);
+static void disconnect_client_from_provisioning_service();
+static void sleep_for_seconds(uint32_t seconds);
+
+int main()
+{
+  create_and_configure_client();
+  LOG_SUCCESS("Client created and configured.");
+
+  connect_client_to_provisioning_service();
+  LOG_SUCCESS("Client connected to \"%s\".", global_provisioning_endpoint_buffer);
+
+  subscribe_client_to_provisioning_service_topics();
+  LOG_SUCCESS("Client subscribed to provisioning service topics.");
+
+  register_client_with_provisioning_service();
+  LOG_SUCCESS("Client registering with provisioning service.");
+
+  receive_registration_status();
+  LOG_SUCCESS("Client received registration status.")
+
+  disconnect_client_from_provisioning_service();
+  LOG_SUCCESS("Client disconnected from provisioning service.");
+
+  return 0;
+}
+
+static void create_and_configure_client()
+{
+  int rc;
+  az_span global_provisioning_endpoint;
+  az_span id_scope;
+  az_span registration_id;
+
+  if (az_failed(
+          rc
+          = read_environment_variables(&global_provisioning_endpoint, &id_scope, &registration_id)))
+  {
+    LOG_ERROR(
+        "Failed to read configuration from environment variables: az_result return code 0x%04x.",
+        rc);
+    exit(rc);
+  }
+
+  if (az_failed(
+          rc = az_iot_provisioning_client_init(
+              &provisioning_client, global_provisioning_endpoint, id_scope, registration_id, NULL)))
+  {
+    LOG_ERROR("Failed to initialize provisioning client: az_result return code 0x%04x.", rc);
+    exit(rc);
+  }
+
+  if (az_failed(
+          rc = az_iot_provisioning_client_get_client_id(
+              &provisioning_client, mqtt_client_id_buffer, sizeof(mqtt_client_id_buffer), NULL)))
+  {
+    LOG_ERROR("Failed to get MQTT client id: az_result return code 0x%04x.", rc);
+    exit(rc);
+  }
+
+  if ((rc = MQTTClient_create(
+           &mqtt_client,
+           (char*)az_span_ptr(global_provisioning_endpoint),
+           mqtt_client_id_buffer,
+           MQTTCLIENT_PERSISTENCE_NONE,
+           NULL))
+      != MQTTCLIENT_SUCCESS)
+  {
+    LOG_ERROR("Failed to create MQTT client: MQTTClient return code %d.", rc);
+    exit(rc);
+  }
+
+  generate_sas_key();
```

Comment:
haha took me a second to not say "check return value". good ol' `exit`

Reply:
i thought it looked funny all by itself.  i tried to make the other local functions that return az_result into void, but it got complicated.  so generate_sas_key() is the loner here...